waku/README.md

```diff
@@ -0,0 +1,287 @@
+|`MessagesRequest{}`|Contains details of a request for historic messages.|
+|`MessagesResponse{}`|Represents a request response sent after processing batch of envelopes.|
+|`MessageStore` `interface`|Implemented by `MemoryMessageStore{}`|
+|`MemoryMessageStore{}`|Represents messages stored in a memory hash table.|
+
+---
+
+### `metrics.go`
+
+[`metrics.go`](./common/metrics.go) is home to [Prometheus](https://prometheus.io/) metric hooks, for counting a range of Waku related metrics.
+
+---
+
+### `protocol.go`
+
+[`protocol.go`](./common/protocol.go) houses the `Peer` and `WakuHost` interfaces.
+
+`Peer` represents a remote Waku client with which the local host waku instance exchanges data / messages. 
+
+`WakuHost` is the local instance of waku, which both interacts with remote clients (peers) and local clients (like `status-react`, via a RPC API).
+
+---
+
+### `rate_limiter.go`
+
+[`rate_limiter.go`](./common/rate_limiter.go) was introduced as an improvement to Whisper allowing Waku nodes to limit the rate at which data is transferred. These limits are defined by the `RateLimits{}` which allows nodes to limit the following:
+
+|Limit Type|Description|
+|---|---|
+|`IPLimits`|Messages per second from a single IP (default 0, no limits)|
+|`PeerIDLimits`|Messages per second from a single peer ID (default 0, no limits)|
+|`TopicLimits`|Messages per second from a single topic (default 0, no limits)|
+
+In addition to the `RateLimits{}` this file also contains the following interfaces and structs.
+
+|Name|Description|
+|---|---|
+|`RateLimiterPeer` `interface`|// TODO|
+|`RateLimiterHandler` `interface`|// TODO|
+|`MetricsRateLimiterHandler{}`|Implements `RateLimiterHandler`, // TODO|
+|`DropPeerRateLimiterHandler{}`|// TODO|
+|`PeerRateLimiterConfig{}`|// TODO|
+|`PeerRateLimiter{}`|// TODO|
+
+---
+
+### `topic.go`
+
+[`topic.go`](./common/topic.go) houses the `TopicType` type.
+
+`TopicType` represents a cryptographically secure, probabilistic partial classification of a message, determined as the first (leftmost) 4 bytes of the SHA3 hash of some arbitrary data given by the original author of a message.
+
+Topics are used to filter incoming messages that the host's user has registered interest in. For further details on filtering see [filter.go](#filtergo).
+
+---
+
+## Versioned
+
+For details about the divergence between versions please consult the `README`s of each version package.
+
+- [version 0](./v0)
+- [version 1](./v1)
+
+---
+
+### Version `const.go`
+
+`const.go` is home to the version sub-package's `const`s. These constants are version dependant and, as expected, may change from version to version.
+
+---
+
+### Version `init.go`
+
+`init.go` is home to the version sub-package's initialisation, and is used to initialise struct based variables at runtime. 
+
+---
+
+### Version `message.go`
+
+`message.go` // TODO - I don't really understand what this does.

```

Comment:
- [ ] Complete this

Reply:
`message.go` is used for what we call message confirmations:

In the UI when a message is sent, it is marked initially as "Pending" and eventually as "Sent".

In order to trigger the transition from pending to "sent", there are these "message confirmations".

Each peer on receiving a message will send back one of these message responses https://github.com/status-im/status-go/blob/4d00656c41909ccdd80a8a77a0982bd66f74d29e/waku/v1/message.go#L31 .

The waku host will check that this is a mailserver and if it's been selected by the user.
If so the message will be marked as "Sent" in the UI.

https://github.com/status-im/specs/blob/master/docs/stable/3-whisper-usage.md has a section on it.

As a side not, we should not be doing any versioning at this level any more (the struct has a `Version` field), something to remove once we decide to move to `waku/2` for example.

Reply:
Resolved with https://github.com/status-im/status-go/pull/1949/commits/bc17559d4284f4d3bbdee51d6dc9dd7a63bacf7e

Reply:
@cammellos

**I'm going to implement the following:**

- Move send bundle to Peer
- change v*/message.go to v*/message_response.go - I know that I originally merged message response functionality into message, but after the `waku/1` refactor it makes sense to move it back.

**I've got some questions:**

- Should `MultiVersionResponse` and `Version1MessageResponse` be merged into a single struct MessageResponse? Or at least renamed to be more obvious what they are.
- Should the structs in v*/message be unexported as they are only consumed inside the version subpackage?
- Should Waku "message response" be renamed as "message confirmation"? It is called a message confirmation in the docs, but also a message response, it makes more sense as confirmation because a message response can be anything that responds to a message including a confirmation, a confirmation is explicitly a confirmation.
- Should MessagesResponse be renamed to MessageResponse
- Should we track new features for the next Waku version? If so where would be a good place?

See :
>Versioning at the MessageResponse level (see the struct Version field) should be phased out and defer to the subpackage's version number. Consider removal once we decide to move to a new major Waku version (i.e. waku/2).

Reply:
> Should MultiVersionResponse and Version1MessageResponse be merged into a single struct MessageResponse? Or at least renamed to be more obvious what they are.

I would say so, if feasible one single struct makes sense.

> Should the structs in v*/message be unexported as they are only consumed inside the version subpackage?

Yes

> Should Waku "message response" be renamed as "message confirmation"?
I'd think so, not sure we use it for anything other than messages confirmations

> Should we track new features for the next Waku version? If so where would be a good place?

I'd think that would be the specs repo, as any new feature should be driven by the specs I take, what do you think?

Reply:
I agree. I was thinking about making the renaming and merging changes now in the open tidy PR, but on reflection I think that we need consensus on the use of terminology and the direction we want to take improvements to the specs.

I will open a spec issue on this subject once I'm done [updating the status specs](https://github.com/status-im/specs/issues/61).